### DISCUSSION TITLE
Downsampling of metric data after a certain period

### DISCUSSION THREAD

**Discussion by ManuelBauer** [grafana/mimir#1834] [Help and support] [answered]
Hello everyone :)

I'm using Grafana Mimir as Prometheus data backend for a few days now and I'm very pleased with its performance.

Now I have a question about data downsampling, because I didn't find anything about it in the documentation. As far as I understand Mimir will store all data for an unlimited time as long as I don't configure a retention time. Is it possible to configure it, that after some time (for example a week) the raw data will be aggregated as (for example) one hour data points which will be represented by a min, max and average value and the corresponding raw data will be deleted?

Would be interesting if something like this is possible, because I don't need the raw data of every metric for all time, but some trend metrics would be nice. Because for reports like "how many http requests hit the loadbalancer two years ago" I don't need every 10s data point 😄.

Thanks in advance for your help :)

Kind regards,
Manuel

**Comment by 56quarters** [comment]
Hello!

No, down sampling is not currently supported. I'm not sure if there are plans to work on it in the future. Are you interested in this because long range queries are taking a long time? Or to save on storage costs?  Or some other reason?

**Reply by ManuelBauer** [reply]
Thanks for the fast answer :)

My main interest is to reduce storage space/cost in the long run (and as a bonus faster queries, because of the reduced dataset 😄).

**Comment by pstibrany** [comment, marked as answer]
Thank you for your question. There are several problems with downsampling that need to be considered:

- when using object store like GCP or S3, storage costs are typically only small fraction of cost of running Mimir, so saving here may not be high.

- downsampling is very IO intensive. Because it's not possible to modify TSDB blocks in place, downsampling requires that blocks are downloaded, rebuilt from scratch with downsampled series (possibly only some of them, based on configuration) and then uploaded back. Old blocks must be deleted. All this processing adds to the cost.

- downsampling complicates querying. PromQL query engine uses single look-back period when looking for samples. Typically this is configured to 5 minutes. Doing downsampling to 1h (with 1 sample per hour) would then cause lot of gaps, if start/end times and step were not carefully selected to work with downsampled data. It might be possible to use longer look-back period, but this is applied "globally" for all series, regardless of which time range is queried (ie. same look-back is used for old data as well as new data). If we wanted to provide different downsampling options to different series, single lookback would not work.

These are just random thoughts that we would need to take into consideration when designing downsampling feature.

**Reply by pracucci** [reply]
Great write up @pstibrany.

Another thing to consider is that in order to provide accurate query results we would have to keep multiple values for 1 downsampled period, like the min or max of a series samples for each downsampled period. There's still opportunity to reduce the number of samples with downsampling (and so reduce storage, and increase query performance) but it's not a reduction from N to 1 per downsampled period.

**Reply by anthonyeleven** [reply]
A block by default is two hours of metrics?  Might one be able to achieve coarse downsampling of historical data by deleting, say, every other block?  Would it be possible to decrease the block size to decrease the blast radius of a deletion?

**Reply by phonkd** [reply]
if users query a big time range (e.g last 90 days) wouldnt downsampling result in way smaller queries (memory and cpu)...?

**Comment by BlueBlue-Lee** [comment]
I don't care the storage cost but care user experience for long time range query. The main goal of downsampling is providing an opportunity to get fast results for long range queries like months or years.

**Comment by wilfriedroset** [comment]
As discussed on slack, I would like to share my usecases
ref: https://grafana.slack.com/archives/C039863E8P7/p1656596260566049

As users, we would like to query a wide range of serie(s). The full resolution is not mandatory. However, when runnning a wide range query the response time impact the user experience.
In this case, downsampling would help to not query all samples and therefore improve reponse time.
Depending on the query, I would like to be able to chose between: min, max, avg, count, sum.
On top on the previous downsampling method, I would like to be able to downsample my series while retaining the visual aspect of the chart. LTTB could be useful in this case.
For one serie we would add 6 more (min, max, avg, count, sum, [LTTB](https://github.com/sveinn-steinarsson/flot-downsample)) as such it is expected to increase the storage cost.
To improve flexibility, as mimir can be used as multi tenant, the downsampling should be configurable at the cluster level (via config) and the tenant level (via runtime config)

```yaml
# cluster level configuration
compactor:
  downsampling:
    - 1d:1m # After 1d apply downsampling add keep 1 sample per minute
    - 2d:5m
    - 2w:1h

# runtime configuration
overrides:
  tenant1:
    downsampling:
      - 1d:1m # After 1d apply downsampling add keep 1 sample per minute
      - 5d:5m
      - 4w:1h
```

We can even think of downsampling differently a subset of the series associated with a tenant based on a regex.

```yaml
overrides:
  tenant1:
    downsampling:
      ".*":
        - 1d:1m # After 1d apply downsampling add keep 1 sample per minute
        - 5d:5m
        - 4w:1h
      "cpu.*":
        - 1d:1m # After 1d apply downsampling add keep 1 sample per minute
        - 4w:1h
```

In some case, once downsampled, the full resolution might not been needed.
As such it could be wishable to be able to define several retention period based on serie name regex. This would help to lower the storage cost.

## Examples
### High frequency sampling

End users want to sample at 1Hz but the full resolution is needed only for a short period of time.
### Capacity planning

Keep low resolution data for capacity planning.

Indeed, the more data we have in the past the more accurate the forecasting is.

The use case could be several years (2 to 5 years).
### Prune full resolution

Only keep the downsampled data after a pre-defined period of time
## Additional context

https://github.com/influxdata/influxdb/issues/23108

## side note

This would help users migrate from other backend without feature loss:

- Thanos https://thanos.io/tip/components/compact.md/#downsampling
- TimescaleDB https://docs.timescale.com/promscale/latest/downsample-data/caggs/#continuous-aggregates-in-promscale
- VictoriaMetrics https://docs.victoriametrics.com/#downsampling (paid feature)
- M3DB https://m3db.io/docs/how_to/m3aggregator/

**Comment by BenoitPoulet** [comment]
Hello,

I must choose a new metric storage and I would like to use downsampling.

Is this planned to add this to Mimir or I must go with Thanos ?

Tky.

**Reply by pracucci** [reply]
Downsampling work hasn't been started in Mimir yet. That being said, could you clarify why you need downsampling? Generally speaking there are two main use cases why people wants downsampling: (1) reduce storage utilization (2) faster queries. In the Mimir context, storage costs are typically low (10% of total cluster cost in my experience) while query acceleration is implemented through other techniques.

**Reply by BenoitPoulet** [reply]
Hello,

the main reasons are :
  - Reduce storage
  - Reduce the number of points to work on when the range is wide. Without that the queries takes ages.

**Reply by 09jvilla** [reply]
@BenoitPoulet -- You'll see that Thanos docs specifically point out that their implementation of downsampling is not for saving on storage costs:
https://thanos.io/tip/components/compact.md/#downsampling

> Keep in mind that the initial goal of downsampling is not saving disk or object storage space. In fact, downsampling doesn’t save you any space but instead, it adds 2 more blocks for each raw block which are only slightly smaller or relatively similar size to raw blocks. This is done by internal downsampling implementation which, to ensure mathematical correctness, holds various aggregations. This means that downsampling can increase the size of your storage a bit (~3x), if you choose to store all resolutions (recommended and enabled by default).

**Reply by ctml91** [reply]
> @BenoitPoulet -- You'll see that Thanos docs specifically point out that their implementation of downsampling is not for saving on storage costs: https://thanos.io/tip/components/compact.md/#downsampling
> 
> > Keep in mind that the initial goal of downsampling is not saving disk or object storage space. In fact, downsampling doesn’t save you any space but instead, it adds 2 more blocks for each raw block which are only slightly smaller or relatively similar size to raw blocks. This is done by internal downsampling implementation which, to ensure mathematical correctness, holds various aggregations. This means that downsampling can increase the size of your storage a bit (~3x), if you choose to store all resolutions (recommended and enabled by default).

How about the second point?

**Reply by 09jvilla** [reply]
I'm assuming @ctml91 that by second point you mean
> Reduce the number of points to work on when the range is wide. Without that the queries takes ages.

If so, I think this is definitely a problem that downsampling can help solve. 

I think what I'm trying to understand is that if Mimir were to implement downsampling in a way similar to Thanos, where downsampling would help run queries over long time periods, but would not help reduce the amount of data in object storage, would that be good enough? Which of these 2 use cases - decreasing storage footprint vs better performance on multi-month/multi-year queries - is more important?

**Reply by sepich** [reply]
>@BenoitPoulet -- You'll see that Thanos docs specifically point out that their implementation of downsampling is not for saving on storage costs:
https://thanos.io/tip/components/compact.md/#downsampling

You are being theoretical. It does not save if you store all 3 resolutions. But in reality you do not have to. Usually you store raw for 2 weeks, 5m for month, and then 1h for ages. Now to the storage "costs", consider this same block at 3 different resolutions:
```
| 01GGCAM3S4ZV9XACSGJNC61X0S | 2022-10-13T00:00:00Z | 2022-10-27T00:00:00Z | 335h59m59.945s | -295h59m59.945s | 625,734     | 9,847,416,532  | 83,365,984  | 7          | false       | location=myo-1,prometheus=it-myo                                                                         | 0s         | compactor |
| 01GGCYJYA3BDDE78R51FP3HVRN | 2022-10-13T00:00:00Z | 2022-10-27T00:00:00Z | 335h59m59.945s | -95h59m59.945s  | 625,734     | 700,010,422    | 5,336,822   | 7          | false       | location=myo-1,prometheus=it-myo                                                                         | 5m0s       | compactor |
| 01GGD5EFRPS2AQ3YH98QHEPVNC | 2022-10-13T00:00:00Z | 2022-10-27T00:00:00Z | 335h59m59.945s | -               | 625,734     | 59,031,659     | 1,135,290   | 7          | false       | location=myo-1,prometheus=it-myo                                                                         | 1h0m0s     | compactor |
```
Their sizes:
```
$ gsutil du -sh gs://thanos/01GGCAM3S4ZV9XACSGJNC61X0S
14.57 GiB    gs://thanos/01GGCAM3S4ZV9XACSGJNC61X0S
$ gsutil du -sh gs://thanos/01GGCYJYA3BDDE78R51FP3HVRN
4.18 GiB     gs://thanos/01GGCYJYA3BDDE78R51FP3HVRN
$ gsutil du -sh gs://thanos/01GGD5EFRPS2AQ3YH98QHEPVNC
560.08 MiB   gs://thanos/01GGD5EFRPS2AQ3YH98QHEPVNC
```
So in thanos it works great.

>In the Mimir context, storage costs are typically low (10% of total cluster cost in my experience)

We have >2Tb in thanos (on GCS) with unlimited retention. Now it is over 2y. And this statement about 10% is totally wrong in our case.

**Reply by 09jvilla** [reply]
> And this statement about 10% is totally wrong in our case.

@sepich can you give us a sense of what fraction of your costs to run Mimir are attributable to object storage if you're saying that the 10% number is way off?

**Reply by sepich** [reply]
We do not run Mimir because it does not support downsampling, we run Thanos. In Thanos you can ship blocks directly to object storage from prometheus (skipping remote-write). In our case most of blocks are being shipped this way, and thanos-receive is a smaller part.
GSC now is 3.48Tb and 3years of data. It is 3480Gb*$0.020=69.6$/mo (no IOPS cost here)
It is only 2 weeks of raw (`15s`) data stored, and the rest is `1h`. 
So if we store 3y of raw blocks - it would be 3.48Tb * 3y * (52w in a year / 2w of raw now) = 271.4Tb = 5428$/mo

Now to compute costs:
sum of thanos-store + thanos-query + thanos-query-frontend + thanos-compact container resources is 2cpu/10Gi
We use gke spot instances, so it is 20$/mo

**Reply by wilfriedroset** [reply]
To give another point of view. I'm running on OVHcloud (full disclosure this is where I work).
My biggest, and oldest, cluster host 1TB of data for ~9months of production data which various scraping interval.
The storage for raw data account for  1010GB*0.025€=25.25€/mo (IOPS included)
The compute account for 9k€

So even with an eventual downsampling feature  I'm still far away from the expected 10% of cost due to storage.

**Reply by 09jvilla** [reply]
Thanks for that datapoint for Mimir, Wilfried. So for you the object storage costs are actually significantly smaller than 10% of TCO. Of course this number will grow as you grow your retention period beyond nine months, but even if you quadruple that to 36 months (3 years), then we'd expect roughly 100 euro/month. That would put you at slightly more than 10% of your total infra spend on Mimir. (Note that our ~10% estimate is based on the expectation of 13 months retention, which is what we default to). 

I'm not as familiar with how Thanos works so I can't chime in on why there is such a discrepancy relative to the Thanos based estimate from @sepich .

**Reply by wilfriedroset** [reply]
Depending on how I scale my cluster, the storage costs may remain around 10%. Both compute and storage cost grow with the consumption of Mimir's tenant.

For exemple, we need to account for 1 more CPU core for every 300k additional active series, same goes for the distributor with the number of sample/sec. Following this logic we need to account for the rest of the compute. This could add up rather quickly.

Second example, with the default retention of 13 months (reported by @09jvilla), if a tenant loads a dashboard for a 90d timerange 6 months ago (e.g: quarter to quarter comparison), we should expect a fair amount of queries which will be splitted and sharded. This weight down on the read path especially on the querier and the store-gateway and therefore the bill.

**Comment by wilfriedroset** [comment]
> I think what I'm trying to understand is that if Mimir were to implement downsampling in a way similar to Thanos, where downsampling would help run queries over long time periods, but would not help reduce the amount of data in object storage, would that be good enough? Which of these 2 use cases - decreasing storage footprint vs better performance on multi-month/multi-year queries - is more important?

I reckon performance on multi-month/multi-year queries is more important than the storage cost. Object storage pricing is generally low compared to the compute pricing.

On alternative to both tackle performance and cost is to have downsampling and different retention per series. This way you could keep the downsampled metrics longer than the raw metrics. This is not possible at the moment as all metrics of a tenant are kept for the same amount of time.

**Comment by BenoitPoulet** [comment]
>On alternative to both tackle performance and cost is to have downsampling and different retention per series. This way you could keep the >downsampled metrics longer than the raw metrics. 

That's exactly what we need. In our case, we don't care about precises metrics over time, we need to see the trends.
Zooming on a range of 10 mins, 8 months ago is not useful for us.

We would like to have high resolutions metrics for a short range like 1 month, medium resolution metrics after that for like 6 months, and low resolution metrics for several years.

**Comment by Otterpohl** [comment]
Is it maybe worth opening an issue for this since there is appetite from the community? i too have this requirement and for the same reasons as those above ^^

**Reply by adirmatzkin** [reply]
Might be interesting for whomever gets to read this - 
A discussion not related entirely to downsampling metrics data - but a workaround I thought of that might be possible for some more use cases of a need for low-res metrics for long-term storage 🙃

https://grafana.slack.com/archives/C01D981PEE5/p1684249509340409?thread_ts=1684142175.298629&cid=C01D981PEE5

**Reply by adirmatzkin** [reply]
(I will describe my idea also here because history gets deleted in the slack channel - tnx to @dimitarvdimitrov that draw my attention)

Basically - the idea is to generate our metrics at the first place in **all** the "resolutions" we want them to be, and then store and keep them for the relevant retention period by leveraging some configuration overrides with Mimir's multi-tenancy feature.

So practically we could do it in different ways, depending on our metrics types and collection mechanism.
Here are 2 simplified examples for doing it:

1. If the metrics collection is **pull-based**: setting duplicated scrape job(s) with a bigger interval could do the work.
2. If the metrics are created from some sort of aggregation and then **pushed**: you could define a bigger time-window for the generation of each metric point (and by doing so - kind of "increase" the interval between data points)
(that was my original need - keeping span-metrics and service-graph metrics generated from OTEL spans for a longer period but still  maintaining reasonable performance)

Then after scraping/creating the metrics with a bigger "gap" between data points, we could -
Send the high-res (short interval) metrics to a tenant with a shorter (let's say a month) retention period,
Send the low-res (longer interval) metrics to a tenant with a longer (let's say 5 years) retention period.

By doing so, we achieve our goal - a tenant that is dedicated for a longer storage retention period, with a **lower resolution** and **better performance** for wide time range queries.

This solution of course comes with the price of (probably more, but those came to my mind first) - 
1. Storing duplicated data for (let's say) the first month
2. More resources needed for duplicated scraping, aggregations (or whatever it will be)
3. It will probably be much more annoying and less "clean" to manage more scraping, data-sources, dashboards configurations, etc.

But still, considering what @pstibrany mentioned in his answer - it will probably be better from price concerns (no need for downsmapling heavy IO work) and from simplicity concerns (due to the complexity of querying data with different lookback-deltas).

I hope this idea could help some of you 🙃
And would love to hear more ideas / thoughts about it !

**Reply by MrFreezeex** [reply]
Oh that's quite interesting I am planning to do this also with mutli tenant but with a slightly different idea. I am still poking around it and not fully tested but I am fairly positive it should work.

So in this scenario we get two mimir tenants per user, for instance for `user1` it would be `user1` and `user1_lts`. They push as usual their metrics to `user1` in whatever scrape interval they want with full control and `user1` get 6 months retention while `user1_lts` get 5 years.

Then to get metrics on the lts tenant we would use the ruler with federated rule group. We first get the active metrics in a CI that runs periodically (essentially `curl -u "user1:xxxx" "${PROMETHEUS_URL}/api/v1/label/__name__/values" -o active-metrics.json`) and then we generate a rule group and push it to mimir (in the lts tenant) that would look like that:
```
groups:
- interval: "1h"
  name: "rules-lts.yaml"
  rules:
  - expr: "metric1"
    record: "metric1"
  - expr: "metric2"
    record: "metric2"
  source_tenants:
  - "user1"
```

In the end we should get virtually 1 scrape every hours in the lts tenant and the normal scrape interval in the other tenant. Users would use the normal tenant for their normal operation and use the lts tenant for things that would involve general trends of the service over the years.

It is in no way perfect but I should works fairly well (hopefully) for most case I believe with the drawback that you loose some data and not effectively aggregate them... And we can both drop many scrape point for lts which is nice for saving storage cost + the query performance with 0 change client side and use the Mimir ruler for everything.

**Reply by adirmatzkin** [reply]
Awesome idea 🔥
Thank you for sharing!

**Reply by wilfriedroset** [reply]
Wow, that's nice to share your tips.
However, I reckon that the downsampling feature is needed. DIY solutions like using the rulers and/or pushing already downsampled data might not be possible (e.g: network devices, IOT).
If a tenant starts to use the rulers to downsampled all of its metrics the induced load will be rather high for the rulers.
Hence my proposal to add a downsampling feature directly in mimir.

**Reply by jrhrmsll** [reply]
Tested the federate approach and works well. It is also possible the do the recording in Prometheus and use a proxy to set the tenant based on metrics labels.

**Reply by Leegin-darknight** [reply]
@adirmatzkin Amazing idea. I have one question about it. Even though I scrape the metrics from the target with a higher scrape interval, the metric name remains the same. So how would we separate that the low-resolution metrics have to be sent to a different tenant when the metric names are the same? Wouldn't it send the same metrics to both tenants?

**Reply by adirmatzkin** [reply]
The point is to set the recording rule for a "long-term-storage tenant" to query the "original" tenant - that way the original and "downsampled" series won't be queried together when using them later.

Have you seen @MrFreezeex's [comment](https://github.com/grafana/mimir/discussions/1834#discussioncomment-5931235)?
I believe you're missing [this](https://grafana.com/docs/mimir/latest/references/architecture/components/ruler/#federated-rule-groups) piece in your puzzle 🙃

**Comment by wilfriedroset** [comment]
I've been thinking about a proposal for a long time. Here is what I've come up.
Please contribute so that we can all together design this feature: https://github.com/grafana/mimir/pull/5028

Be advised it is not as easy as it looks.

**Reply by marcusteixeira** [reply]
Hi @wilfriedroset , any update this feature?

**Reply by wilfriedroset** [reply]
Unfortunately no

**Comment by anthonyeleven** [comment]
This is near and dear to my heart, so I left a few suggestions to clarify the excellent document.

**Comment by phonkd** [comment]
imo the argument "s3 is cheap" is invalid, more samples = more expensive queries over big time range --> more read path resources (memory and cpu)

**Comment by Moemad-711** [comment]
Is there any update on the proposal? I think it's already a good proposal and the 1st proposal to use the compactor is a really good idea and better than adding yet another microservice.
